Libvirt/libvirt node driver cleanup and tests

Katana-Steel opened the pull request:
## Extra tests for libvirt_driver and cleaner driver creation
### Description

Code clean up and added more tests for the libvirt_driver module.
Cleaner driver creation with less console output when using SASL
authentication.
### Status
- done, ready for review
### Checklist (tick everything that applies)
- [X] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [X] Documentation
- [X] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)
- [ ] [ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes) (required for bigger changes)

Commit:
cleaned up LibvirtNodeDriver.__init__

checking the uri string for +tcp before trying to do the SASL
connection to the libvirtd service

Commit:
adding more tests, and cleaning up the mock setup

Kami commented on libcloud/test/compute/test_libvirt_driver.py:
```diff
@@ -26,11 +28,19 @@ class virConnect:
+        return []
+
     def __init__(self):
-        stub = self.stub
```
I'm fine with this code, but I believe you could also utilize Mock for that by specifying a class you want to mock for `spec` argument - http://www.voidspace.org.uk/python/mock/mock.html#mock.Mock.

This would mean less code to maintain and keep up to date :)

Kami commented on libcloud/test/compute/test_libvirt_driver.py:
```diff
@@ -65,103 +75,66 @@ def __init__(self):
+    name = 'Libvirt'
+    website = 'http://libvirt.org/'
+
+    NODE_STATE_MAP = {
```
Hm, do we need this attributes here on the test class? They are already defined in the `LibvirtNodeDriver` class.

Also, usually the idea is not to have the test class inherit from the Driver class, but the test case should instantiate the driver class and use that.

For example:

``` python
def setUp(self):
    ....
    self.driver = LibvirtNodeDriver(...)

def some_test_method(self):
    nodes = self.driver.list_nodes()
    ...
```

Kami replied:
It would be great if you could make those changes so it's consistent with other drivers tests and it's a bit easier to maintain it - thanks!

Kami commented on libcloud/compute/drivers/libvirt_driver.py:
```diff
@@ -82,16 +82,18 @@ def __init__(self, uri, key=None, secret=None):
                     self._cred_callback, None]
             self.connection = libvirt.openAuth(uri, auth, 0)
+        else:
+            self.connection = libvirt.open(uri)
```
Hm, so `open` simply returns `None` if it's unable to establish a connection?

If so, that's a bit unfriendly and I guess if you want to debug it you need to set `LIBVIR_DEBUG` environment variable, or is there any other, more user-friendly way to do it?

Katana-Steel replied:
my bad... actually it will raise libvirt.libvirtError in case of a failed connection, fixing

Kami commented:
Thanks for this contribution.

I've added some comments. It would be great if you could at least update the tests to follow the same approach as other driver test (sorry for not mentioning in your first tests PR, I missed it :().

Commit:
replaced inheited class with autospec'ed mock

Commit:
passing None to open will use defaults

if we get a connection with None, retrive the connection uri from libvirt.virConnection

Katana-Steel commented:
reviewed the suggestions and I did read up on Mock for testing
and updated the test code to mock essentially the entire libvirt module just to be safe ;)

Commit:
import mock from the correct unittest module

Commit:
installing mock from pip

Commit:
require tox to include Mock modules

Commit:
import mock module directly

the mock module is not in unittest2 so include it directly

Commit:
adding mock to more 2.x python testing

Katana-Steel commented:
the have_libvirt check makes using the mock framework hard when libvirt isn't installed.

I'll take any suggestions... I could go back to inherit and just use MagicMock for virConnection

it seems bad to assume libvirt is installed for tests to pass. Or would that be an acceptable change?

Commit:
adding python libvirt to dependencies

Kami commented on tox.ini:
```diff
@@ -37,10 +38,12 @@ deps = -r{toxinidir}/requirements-tests.txt
 [testenv:py2.6-lxml]
 deps = -r{toxinidir}/requirements-tests.txt
        unittest2
+       mock
```
mock is already specified in requirements-tests.txt so no need to also specify it here.

In any case, not a big deal, I will remove it while merging :)

Kami commented:
@Katana-Steel Ideally we can make it work without requiring libvirt to be installed, at least for the basic unit tests.

I'm fine with having additional integration tests which can require libvirt to be installed as long as those don't run by default if libvirt is not installed (we can use unittest.skipIf and skip if libvirt is not installed).

Kami commented:
Thanks for making those changes.

I will make the mock change and get the tests passing locally and then merge it into trunk.

Kami commented:
I made those changes and merged it into trunk - 01c6db45e532958d114033ab1b8e271351bf1c26 , d9fc1edaf6b3f52b1141ab80c1d88d97a964380a, a4bddfec0b26606258f0e1da94c289d0ab573369

Thanks!

Kami commented:
Also - 6626808f2e301685b88a7cf3cbe3a13d14ef78ce

Some time in the next couple of days when I get a chance I will look into running libvirt tests on Travis CI.

For one, we need to make sure all the system level libvirt dependencies are installed on Travis and after that we need to install `libvirt` library in all the virtual environments used by tox.

Kami commented:
It looks like that Travis doesn't have libvirt system package whitelisted so sadly it doesn't look like we can get this easily to run on new Travis container based infrastructure :/

https://travis-ci.org/apache/libcloud/jobs/144437976#L118
